Client-Server/Server/SeTCPServer.m

```diff
@@ -0,0 +1,88 @@
+#import "SeTCPServer.h"
+#import <Foundation/Foundation.h>
+#import <CoreFoundation/CoreFoundation.h>
+#import <sys/socket.h>
+#import <netinet/in.h>
+
+@interface SeTCPServer()
+
+@property (nonatomic) CFSocketRef socket;
+@property (nonatomic) NSInputStream *inputStream;
+
+@end
+
+@implementation SeTCPServer
+
+- (id)init {
+    return [self initWithPort:6666];
+}
+
+- (id)initWithPort:(NSUInteger)port {
+    if(([super init])) {
+        _port = port;
+    }
+
+    return self;
+}
+
+- (void)start {
+    self.socket = CFSocketCreate(kCFAllocatorDefault, PF_INET, SOCK_STREAM, IPPROTO_TCP, kCFSocketAcceptCallBack, serverConnectCallback, nil);
+   
+    if(!self.socket) {
+        NSLog(@"CFSocketCreate failed");
+        return;
+    }
+
+    struct sockaddr_in sin = {0};
+    sin.sin_len = sizeof(sin);
+    sin.sin_family = AF_INET; /* Address family */
+    sin.sin_port = htons(self.port); /* Or a specific port */
+    sin.sin_addr.s_addr = INADDR_ANY;
+
+    CFDataRef sinData = CFDataCreate(kCFAllocatorDefault, (UInt8 *)&sin, sizeof(sin));
+    if (CFSocketSetAddress(self.socket, sinData)) {
+        NSLog(@"CFSocketSetAddress failed");
+        return;
+    }
+
+    CFRunLoopSourceRef socketsource = CFSocketCreateRunLoopSource(kCFAllocatorDefault, self.socket, 0);
+    CFRunLoopAddSource(CFRunLoopGetCurrent(), socketsource, kCFRunLoopDefaultMode);
+    CFRunLoopRun();
+}
+
+// socket connection w/ client side. create another data source and add to run-loop
+void serverConnectCallback(CFSocketRef s, CFSocketCallBackType callbackType, CFDataRef address, const void *data, void *info) {
+    CFSocketContext socketContext;
+    memset(&socketContext, 0, sizeof(CFSocketContext));
+    int clientfd = *((int *)data);
+    socketContext.info = (void *)((long)clientfd);
+    // create CFSocket for w/ connected client
+    CFSocketRef socket = CFSocketCreateWithNative(kCFAllocatorDefault, clientfd, kCFSocketDataCallBack, clientDataCallback, &socketContext);
+    CFSocketDisableCallBacks(socket, kCFSocketWriteCallBack);
+    CFRunLoopSourceRef socketRunLoop = CFSocketCreateRunLoopSource(kCFAllocatorDefault, socket, 0);
+    CFRunLoopAddSource(CFRunLoopGetCurrent(), socketRunLoop, kCFRunLoopCommonModes);
+    CFRelease(socket);
+    CFRelease(socketRunLoop);
+}
+
+// data to/from client
+void clientDataCallback(CFSocketRef s, CFSocketCallBackType callbackType, CFDataRef address, const void *data, void *info) {
+    if (!(callbackType & kCFSocketDataCallBack))  return;
```

Comment:
Since the callback can only be called with a single flag (you can't both have a "connected" and "data" callback at the same time, those are two separate events), it'd be nicer to check via comparison (==)